client/gutenberg/extensions/presets/jetpack/utils/jetpack-block-type.js

```diff
@@ -0,0 +1,43 @@
+/**
+ * External dependencies
+ */
+import { registerBlockType } from '@wordpress/blocks';
+import { every, has, get, pickBy } from 'lodash';
+
+/**
+ * Where Jetpack data can be found in WP Admin
+ */
+const JETPACK_DATA_PATH = [ 'Jetpack_Initial_State' ];
+
+export default class JetpackBlockType {
+	constructor( name, config ) {
+		this.name = name;
+		this.config = config;
+		this.requiredModules = null;
+		// TODO: how do we access Jetpack Data from calypso?
+		this.jetpackData = get(
```

Comment:
Seems like we can access the information "whether to register or not" just once, at the time when we instantiate each `JetpackBlockType`. This has some downsides - we always need to have this information available, and in Calypso we'll probably need to request it first. This means, we need it to work asynchronously in some envs.

So, how about if we create a custom `wp.data` store for Jetpack block visibility, and have a `withSelect()` with that store? For that store:

* For Jetpack, we could pass it part of `Jetpack_Initial_State` as an initial state. 
* For WP.com / Calypso (in other words, when the global var `Jetpack_Initial_State` is missing), we could pass it an empty, generic default state. 

Then, using `wp.data`'s default behavior of how selectors and resolvers work, we could:

* For Jetpack, already having the block visibility information, proceed to register the blocks immediately.
* For WP.com / Calypso, asynchronously start loading the block visibility information, and then register the blocks after it loads.

To account for any differences between how we perform API requests between environments, we could make use of the custom middleware support of `apiFetch`.

With this approach, as an end result:

* For Jetpack, blocks would be registered synchronously.
* For WP.com/Calypso, blocks would be registered asynchronously.